### PR TITLE
ChibiOS: enable UART FIFO on H7 and G4 by default

### PR DESCRIPTION
**Pull request by tridge**
This enables the hardware FIFO on UARTs on G4 and H7 by default. It can be disabled using a SERIALn_OPTIONS bit.
Enabling the fifo makes it possible to do very high baudrates without DMA. Testing on a CubeOrange shows no byte loss at 2MBit/s between telem1 and telem2 without DMA and with the FIFO enabled.
This makes moving baseline dual GPS yaw more reliable on H7 as it removes DMA issues for the RTCM data on H7 based boards. Testing with two F9P u-blox modules shows reliable yaw with DMA disabled on both ports. Disabling DMA actually helps as it stops contention with other devices on the same DMA channels, lowering latency for the RTCM data.
Setting SERIALn_OPTIONS to 768 disables DMA in both directions, which works very well for GPS yaw

### CODE REVIEW

**Comment by amilcarlucas** [comment]
Does this bypass the ring-buffer and writes directly to the UART hardware FIFO?

**Comment by tridge** [comment]
> Does this bypass the ring-buffer and writes directly to the UART hardware FIFO?

no, ring buffer is still there. The big advantage is for receive with no DMA, where we become much less sensitive to interrupt latency and so can receive high baudrates with no loss of bytes without using DMA

**Comment by rmackay9** [comment]
Two experienced beta testers (Yuri and Kenny) have offered to test this over the next week or so.  [See discussion here](https://discuss.ardupilot.org/t/yuris-4-1-0-beta-update-gps-yaw-now-beta3/71100/111).

**Comment by yuri-rage** [comment]
> Two experienced beta testers (Yuri and Kenny) have offered to test this over the next week or so. [See discussion here](https://discuss.ardupilot.org/t/yuris-4-1-0-beta-update-gps-yaw-now-beta3/71100/111).

I started testing today and noticed no real improvement so far.  Autopilot-in-the-middle communication between moving base and rover has never been very reliable for me (Cube Orange, SERIAL3 and SERIAL4).  I'm open to any suggestions for parameter changes beyond those described in the discussion that Randy linked.  I can provide logs for any tested configuration.

**Comment by tridge** [comment]
>  I can provide logs for any tested configuration.

I'd like the log for both GPS ports set to 768 options for no DMA, thanks!

File: libraries/AP_SerialManager/AP_SerialManager.cpp [hunk 1]
```diff
@@ -569,7 +569,13 @@ AP_HAL::UARTDriver *AP_SerialManager::find_serial(enum SerialProtocol protocol,
         return nullptr;
     }
     const uint8_t serial_idx = _state - &state[0];
-    return hal.serial(serial_idx);
+
+    // set options before any user does begin()
+    AP_HAL::UARTDriver *port = hal.serial(serial_idx);
+    if (port) {
+        port->set_options(_state->options);
```
Comment (bugobliterator) [inline libraries/AP_SerialManager/AP_SerialManager.cpp:576]: wouldn't it be better to have this in begin method itself, basically UART driver fetches the options?

Comment (tridge) [reply libraries/AP_SerialManager/AP_SerialManager.cpp:576]: set_options() is also used to changes options from places other than the parameters, so if we always fetched from the parameters we'd get the wrong value. See calls like this:
  libraries/AP_RCProtocol/AP_RCProtocol.cpp:            added.uart->set_options(added.uart->get_options() | AP_HAL::UARTDriver::OPTION_RXINV);

**Review by rmackay9 — APPROVED** [review]
This has been tested by Yuri the rover user so I'm approving, thanks!

**Pull request merged**